Mathlib/Lean/Expr.lean

```diff
@@ -61,6 +62,74 @@ def natLit! : Expr → Nat
 
+
+/-! We define a more flexible version of `Expr.replace` where we can use recursive calls even when
+  replacing a subexpression. We completely mimic the implementation of `Expr.replace`. -/
+namespace ReplaceRecImpl
+
+abbrev cacheSize : USize := 8192
+
+structure State where
+  -- Remark: our "unsafe" implementation relies on the fact that `()` is not a valid Expr
+  keys    : Array Expr
+  results : Array Expr
+
+abbrev ReplaceM := StateM State
+
+@[inline] unsafe def cache (i : USize) (key : Expr) (result : Expr) : ReplaceM Expr := do
+  modify fun ⟨keys, results⟩ => { keys := keys.uset i key lcProof, results := results.uset i result lcProof };
+  pure result
+
+@[inline] unsafe def replaceUnsafeM (f? : Expr → Option (Array Expr × (Array Expr → Expr)))
+  (size : USize) (e : Expr) : ReplaceM Expr := do
+  let rec @[specialize] visit (e : Expr) : ReplaceM Expr := do
+    let c ← get
+    let h := ptrAddrUnsafe e
+    let i := h % size
+    if ptrAddrUnsafe (c.keys.uget i lcProof) == h then
+      pure <| c.results.uget i lcProof
+    else match f? e with
+      | some (es, g) => do
+        let esNew ← es.mapM visit
+        cache i e (g esNew)
+      | none      => match e with
+        | Expr.forallE _ d b _   => cache i e <| e.updateForallE! (← visit d) (← visit b)
+        | Expr.lam _ d b _       => cache i e <| e.updateLambdaE! (← visit d) (← visit b)
+        | Expr.mdata _ b _       => cache i e <| e.updateMData! (← visit b)
+        | Expr.letE _ t v b _    => cache i e <| e.updateLet! (← visit t) (← visit v) (← visit b)
+        | Expr.app f a _         => cache i e <| e.updateApp! (← visit f) (← visit a)
+        | Expr.proj _ _ b _      => cache i e <| e.updateProj! (← visit b)
+        | e                      => pure e
+  visit e
+
+unsafe def initCache : State :=
+  { keys    := mkArray cacheSize.toNat (cast lcProof ()), -- `()` is not a valid `Expr`
+    results := mkArray cacheSize.toNat arbitrary }
+
+@[inline] unsafe def replaceUnsafe (f? : Expr → Option (Array Expr × (Array Expr → Expr)))
+  (e : Expr) : Expr :=
+  (replaceUnsafeM f? cacheSize e).run' initCache
+
+end ReplaceRecImpl
+
+/-- A version of `Expr.replace` where we can use recursive calls even if we replace a subexpression.
+  When reaching a subexpression `e` we call `f? e` to see if we want to do anything with this
+  expression. If `f? e = none` we proceed to the children of `e`. If
+  `f? e = some (#[e₁, ..., eₙ], g)`, we first recursively apply this function to
+  `#[e₁, ..., eₙ]` to get new expressions `#[f₁, ..., fₙ]`.
+  Then we replace `e` by `g [f₁, ..., fₙ]`.
+
+  Important: In order for this function to terminate, the `[e₁, ..., eₙ]` must all be smaller than
+  `e` according to some measure  (and this measure must also be strictly decreasing on the w.r.t.
+  the structural subterm relation). -/
+@[implementedBy ReplaceRecImpl.replaceUnsafe]
+partial def replaceRec (f? : Expr → Option (Array Expr × (Array Expr → Expr))) (e : Expr) :
```

Comment:
```suggestion
partial def replaceRec (f? : (Expr → Expr) → Expr → Option Expr) (e : Expr) :
```
I wonder if that isn't a more comfortable inferface (implemented with IO.Ref instead of StateM).

Reply:
That would be a lot nicer interface, but I don't see how to program this. 
* I don't know how to use `IO.Ref` here. Are you suggesting the monad to be `EIO State`? What `IO`-function will be helpful here that `StateM` doesn't provide?
* I'd be also happy with `(f? : Expr → Option ((Expr → Expr) → Expr))`, which might make it easier to implement?
* If `f?` has that type, I can sort-of do it with `StateM`, when replacing the relevant lines to
```lean
    else match f? e with
      | some g => do 
        let s ← get
        let eNew := g λ e => ((visit e).run s).1
        cache i e eNew
```
This has the obvious problem that I don't update the state when computing `eNew`. Can I do it better with `IO`?

Reply:
I was thinking of something like this (but replace the cache type):
```lean
import Std
import Mathlib.Util.TermUnsafe
open Std ShareCommon

unsafe def memoFixImpl {α β : Type} [Inhabited β] (f : (α → β) → (α → β)) (a : α) : β := unsafeBaseIO do
  let cache : IO.Ref (@HashMap Object β ⟨Object.ptrEq⟩ ⟨Object.ptrHash⟩) ← ST.mkRef {}
  let rec fix (a : α) : β := unsafeBaseIO do
    if let some b := (← cache.get).find? (unsafeCast a) then
      return b
    let b := f fix a
    cache.modify (·.insert (unsafeCast a) b)
    b
  fix a

@[implementedBy memoFixImpl]
constant memoFix {α β : Type} [Inhabited β] (f : (α → β) → (α → β)) : α → β

def fibLength : List Nat → Nat :=
  memoFix fun fib xs =>
    dbg_trace xs
    match xs with
    | [] => 0
    | [x] => 1
    | _ :: xs@(_ :: ys) => fib xs + fib ys

-- visits every sublist only once
#eval fibLength [1,2,3,4,5,6,7,8]
```

Reply:
Ok, I'll try using this. 
I was also wondering about the choice of cache in `Expr.replace`. Obviously `Array` is very quick, but pointer collisions (mod `cacheSize`) seem very expensive, so I was wondering whether a `HashMap`/`RBTree` would be better.

Reply:
> Obviously Array is very quick, but pointer collisions (mod cacheSize) seem very expensive, so I was wondering whether a HashMap/RBTree would be better.

Good question!  I don't really know.  I would imagine that both implementation are good enough for toAdditive, though.

Reply:
I've now implemeted Gabriel's suggested signature using a memoFix